sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -17,6 +18,52 @@
 from ._bitset import set_bitset_memoryview
 
 
+def _pack_matrix(sparse_matrix):
+    """Pack a sparse matrix into a dense one.
+
+    Packing is done column-wise.
+
+    Parameters
+    ----------
+    sparse_matrix : csr_matrix, shape (n_samples, n_features)
+
+    Return
+    ------
+    dense_matrix : ndarray of shape(n_non_zero_samples, n_features)
+        The sparse_matrix packed, the number of rows is equal to
+        the biggest number of non-zero samples per column.
+    row_indices : ndarray of shape(n_non_zero_samples, n_features)
+        The initial row indice of the packed values
+    """
+    columns = [[] for _ in range(sparse_matrix.shape[1])]
+    indices = [[] for _ in range(sparse_matrix.shape[1])]
+    row_indices, column_indices, values = sp.find(sparse_matrix)
+    for i in range(len(values)):
+        # sp.find return indices sorted by column so we can just
+        # append to the list
+        columns[column_indices[i]].append(values[i])
+        indices[column_indices[i]].append(row_indices[i])
+    n_rows = max(map(lambda x: len(x), columns))
+    # Fill the missing values
+    for j in range(len(columns)):
+        columns[j].extend([0 for _ in range(n_rows-len(columns[j]))])
+        indices[j].extend([np.nan for _ in range(n_rows-len(indices[j]))])
+    return np.array(columns, dtype=sparse_matrix.dtype).T, np.array(indices).T
+
+
+def _unpack_matrix(dense_matrix, row_indices, shape):
+    sparse_matrix = sp.lil_matrix(shape, dtype=dense_matrix.dtype)
```

Comment:
From my experience, using `lil_matrix` to construct medium size matrices and then converting to a `tocsr` is not great time wise and memory wise.

If we are building a binned matrix, we know for sure that it will have the same nonzero elements as the input sparse matrix. We can construct a the output by copying over the indptr and indices, and setting the data to be the same shape but with the binned dtype.